## XỬ LÍ - LÀM SẠCH DỮ LIỆU


#### I. Mô tả dữ liệu 

In [83]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
import numpy as np
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import train_test_split


In [84]:
# Đọc dữ liệu từ file csv
stroke_prediction_df = pd.read_csv('File dữ liệu thô.csv')

stroke_prediction_df.head(10)

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   
5  56669    Male  81.0             0              0          Yes   
6  53882    Male  74.0             1              1          Yes   
7  10434  Female  69.0             0              0           No   
8  27419  Female  59.0             0              0          Yes   
9  60491  Female  78.0             0              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   
5        Private          Urban             186.21  29.0  formerly smoked   
6        Private          Rural              70.09  27.4     never smoked   
7        Private          Urban              94.39  22.8     never smoked   
8        Private          Rural              76.15   NaN          Unknown   
9        Private          Urban              58.57  24.2          Unknown   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1  
5       1  
6       1  
7       1  
8       1  
9       1

In [85]:
# Kiểm tra thông tin sơ bộ của Data Frame
stroke_prediction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


Dữ liệu được lấy từ bộ dữ liệu chăm sóc sức khỏe (Healthcare Dataset - Stroke Data), gồm nhiều thông tin liên quan đến hồ sơ y tế của bệnh nhân.
Một số cột quan trọng trong tập dữ liệu gồm:

- id: Mã định danh bệnh nhân (không dùng trong dự đoán).
- gender: Giới tính.
- age: Tuổi.
- hypertension: Tình trạng huyết áp cao.
- heart_disease: Bệnh tim.
- ever_married: Tình trạng hôn nhân.
- work_type: Loại công việc.
- Residence_type: Khu vực sống (nông thôn/thành thị).
- avg_glucose_level: Mức đường huyết trung bình.
- bmi: Chỉ số khối cơ thể.
- smoking_status: Tình trạng hút thuốc.
stroke: Cột mục tiêu (1 nếu có đột quỵ, 0 nếu không).

#### II. Thực hiện tiền xử lí

1. Xử lý Dữ liệu thiếu (Handling Missing Values):

In [86]:

# Xử lý giá trị thiếu ở cột BMI
stroke_prediction_df['bmi'].fillna(stroke_prediction_df['bmi'].median(), inplace=True)

stroke_prediction_df.head(10)



C:\Users\hary0\AppData\Local\Temp\ipykernel_6460\2100059971.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  stroke_prediction_df['bmi'].fillna(stroke_prediction_df['bmi'].median(), inplace=True)


id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   
5  56669    Male  81.0             0              0          Yes   
6  53882    Male  74.0             1              1          Yes   
7  10434  Female  69.0             0              0           No   
8  27419  Female  59.0             0              0          Yes   
9  60491  Female  78.0             0              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21  28.1     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   
5        Private          Urban             186.21  29.0  formerly smoked   
6        Private          Rural              70.09  27.4     never smoked   
7        Private          Urban              94.39  22.8     never smoked   
8        Private          Rural              76.15  28.1          Unknown   
9        Private          Urban              58.57  24.2          Unknown   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1  
5       1  
6       1  
7       1  
8       1  
9       1

- Vì BMI là dữ liệu số, thường phân bố lệch nhẹ, nên median (trung vị) sẽ an toàn hơn mean (trung bình) khi có outlier.

2. Mã hóa dữ liệu (Encoding Categorical Data):

In [87]:
# Các cột cần mã hóa
categorical_cols = ['gender', 'ever_married', 'work_type', 'Residence_type', 'smoking_status']

# Tạo DataFrame chứa các cột đã mã hóa
encoded = pd.get_dummies(stroke_prediction_df[categorical_cols], drop_first=True)

# Ghép lại vào stroke_prediction_df gốc, giữ nguyên thứ tự các cột ban đầu
stroke_prediction_df = pd.concat([stroke_prediction_df.drop(columns=categorical_cols), encoded], axis=1)

stroke_prediction_df.head(10)


id   age  hypertension  heart_disease  avg_glucose_level   bmi  stroke  \
0   9046  67.0             0              1             228.69  36.6       1   
1  51676  61.0             0              0             202.21  28.1       1   
2  31112  80.0             0              1             105.92  32.5       1   
3  60182  49.0             0              0             171.23  34.4       1   
4   1665  79.0             1              0             174.12  24.0       1   
5  56669  81.0             0              0             186.21  29.0       1   
6  53882  74.0             1              1              70.09  27.4       1   
7  10434  69.0             0              0              94.39  22.8       1   
8  27419  59.0             0              0              76.15  28.1       1   
9  60491  78.0             0              0              58.57  24.2       1   

   gender_Male  gender_Other  ever_married_Yes  work_type_Never_worked  \
0         True         False              True                   False   
1        False         False              True                   False   
2         True         False              True                   False   
3        False         False              True                   False   
4        False         False              True                   False   
5         True         False              True                   False   
6         True         False              True                   False   
7        False         False             False                   False   
8        False         False              True                   False   
9        False         False              True                   False   

   work_type_Private  work_type_Self-employed  work_type_children  \
0               True                    False               False   
1              False                     True               False   
2               True                    False               False   
3               True                    False               False   
4              False                     True               False   
5               True                    False               False   
6               True                    False               False   
7               True                    False               False   
8               True                    False               False   
9               True                    False               False   

   Residence_type_Urban  smoking_status_formerly smoked  \
0                  True                            True   
1                 False                           False   
2                 False                           False   
3                  True                           False   
4                 False                           False   
5                  True                            True   
6                 False                           False   
7                  True                           False   
8                 False                           False   
9                  True                           False   

   smoking_status_never smoked  smoking_status_smokes  
0                        False                  False  
1                         True                  False  
2                         True                  False  
3                        False                   True  
4                         True                  False  
5                        False                  False  
6                         True                  False  
7                         True                  False  
8                        False                  False  
9                        False                  False

- Áp dụng phương pháp One-Hot Encoding để chuyển các giá trị phân loại thành các biến nhị phân (0 và 1). Mỗi giá trị trong một cột được mã hóa thành một cột mới đại diện riêng biệt.
Việc này giúp mô hình hiểu đúng bản chất của dữ liệu, tránh việc hiểu sai thứ tự giữa các giá trị (như khi dùng Label Encoding), đồng thời cải thiện khả năng học và dự đoán của mô hình.

3. Xử lý dữ liệu không liên quan (Handling Irrelevant Features):

In [88]:
# Xóa cột id không liên quan
stroke_prediction_df = stroke_prediction_df.drop( columns= 'id',axis =1, inplace = False)

- Trong bộ dữ liệu, cột id chỉ đóng vai trò là mã số định danh cho từng bản ghi (record). Nó không mang thông tin thực chất nào về đặc điểm của bệnh nhân hoặc kết quả bệnh lý, nên không có giá trị dự đoán trong mô hình học máy.

4. Xử lý dữ liệu ngoại lai (Outlier)

In [89]:
numeric_cols = ['age', 'avg_glucose_level', 'bmi']

iqr_factor = 0.5  # giảm từ 1.5 xuống để clip mạnh hơn

for col in numeric_cols:
    Q1 = stroke_prediction_df[col].quantile(0.25)
    Q3 = stroke_prediction_df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - iqr_factor * IQR
    upper_bound = Q3 + iqr_factor * IQR
    stroke_prediction_df[col] = np.clip(stroke_prediction_df[col], lower_bound, upper_bound)

    print(f"{col}: min={stroke_prediction_df[col].min()}, max={stroke_prediction_df[col].max()}")


age: min=7.0, max=79.0
avg_glucose_level: min=58.822500000000005, max=132.5125
bmi: min=19.300000000000004, max=37.3


In [90]:
stroke_prediction_df.head(10)

age  hypertension  heart_disease  avg_glucose_level   bmi  stroke  \
0  67.0             0              1           132.5125  36.6       1   
1  61.0             0              0           132.5125  28.1       1   
2  79.0             0              1           105.9200  32.5       1   
3  49.0             0              0           132.5125  34.4       1   
4  79.0             1              0           132.5125  24.0       1   
5  79.0             0              0           132.5125  29.0       1   
6  74.0             1              1            70.0900  27.4       1   
7  69.0             0              0            94.3900  22.8       1   
8  59.0             0              0            76.1500  28.1       1   
9  78.0             0              0            58.8225  24.2       1   

   gender_Male  gender_Other  ever_married_Yes  work_type_Never_worked  \
0         True         False              True                   False   
1        False         False              True                   False   
2         True         False              True                   False   
3        False         False              True                   False   
4        False         False              True                   False   
5         True         False              True                   False   
6         True         False              True                   False   
7        False         False             False                   False   
8        False         False              True                   False   
9        False         False              True                   False   

   work_type_Private  work_type_Self-employed  work_type_children  \
0               True                    False               False   
1              False                     True               False   
2               True                    False               False   
3               True                    False               False   
4              False                     True               False   
5               True                    False               False   
6               True                    False               False   
7               True                    False               False   
8               True                    False               False   
9               True                    False               False   

   Residence_type_Urban  smoking_status_formerly smoked  \
0                  True                            True   
1                 False                           False   
2                 False                           False   
3                  True                           False   
4                 False                           False   
5                  True                            True   
6                 False                           False   
7                  True                           False   
8                 False                           False   
9                  True                           False   

   smoking_status_never smoked  smoking_status_smokes  
0                        False                  False  
1                         True                  False  
2                         True                  False  
3                        False                   True  
4                         True                  False  
5                        False                  False  
6                         True                  False  
7                         True                  False  
8                        False                  False  
9                        False                  False

- Xử lý giá trị ngoại lai (outlier) giúp mô hình học máy hoạt động ổn định và chính xác hơn. Vì các giá trị quá lớn hoặc quá nhỏ có thể làm sai lệch trung bình, phương sai và hướng học của mô hình, dẫn đến kết quả dự đoán kém tin cậy.

#### III. Cân bằng dữ liệu (Đức Tỉnh)

1. Kiểm tra tỉ lệ của biến mục tiêu

In [91]:
stroke_prediction_df['stroke'].value_counts()

stroke
0    4861
1     249
Name: count, dtype: int64

- Dữ liệu bị mất cân bằng

2. Cân bằng bằng phương pháp SMOTE (Synthetic Minority Oversampling Technique)

In [92]:
# Tách dữ liệu
X = stroke_prediction_df.drop('stroke', axis=1)
y = stroke_prediction_df['stroke']

# Chia dữ liệu 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Cân bằng dữ liệu 
smote = SMOTE(random_state=42)
X_train_res, y_train_res = smote.fit_resample(X_train, y_train)


- Trong tập dữ liệu này, số lượng người không bị đột quỵ (stroke = 0) nhiều hơn rất nhiều so với người bị đột quỵ (stroke = 1).
Nếu không cân bằng, mô hình sẽ có xu hướng dự đoán nghiêng về lớp chiếm đa số, dẫn đến độ chính xác ảo nhưng không phát hiện được các trường hợp đột quỵ thật.

In [93]:


# Chuyển X_train_res thành DataFrame
X_train_res_df = pd.DataFrame(X_train_res, columns=X_train.columns)

# Kết hợp với y_train_res thành 1 DataFrame hoàn chỉnh
stroke_prediction_df_cleand = pd.concat([X_train_res_df, pd.Series(y_train_res, name='stroke')], axis=1)

stroke_prediction_df_cleand.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7778 entries, 0 to 7777
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   age                             7778 non-null   float64
 1   hypertension                    7778 non-null   int64  
 2   heart_disease                   7778 non-null   int64  
 3   avg_glucose_level               7778 non-null   float64
 4   bmi                             7778 non-null   float64
 5   gender_Male                     7778 non-null   bool   
 6   gender_Other                    7778 non-null   bool   
 7   ever_married_Yes                7778 non-null   bool   
 8   work_type_Never_worked          7778 non-null   bool   
 9   work_type_Private               7778 non-null   bool   
 10  work_type_Self-employed         7778 non-null   bool   
 11  work_type_children              7778 non-null   bool   
 12  Residence_type_Urban            77

In [94]:
# Kiểm tra xem dữ liệu đã cân bằng hay chưa
stroke_prediction_df_cleand['stroke'].value_counts() 

stroke
0    3889
1    3889
Name: count, dtype: int64

3. Thống kê mô tả

In [95]:
stroke_prediction_df_cleand.describe()

age  hypertension  heart_disease  avg_glucose_level  \
count  7778.000000   7778.000000    7778.000000        7778.000000   
mean     55.019436      0.092826       0.050913          99.858664   
std      21.323947      0.290207       0.219834          25.453910   
min       7.000000      0.000000       0.000000          58.822500   
25%      43.000000      0.000000       0.000000          77.742688   
50%      58.938779      0.000000       0.000000          96.121378   
75%      74.000000      0.000000       0.000000         132.512500   
max      79.000000      1.000000       1.000000         132.512500   

               bmi       stroke  
count  7778.000000  7778.000000  
mean     28.935360     0.500000  
std       5.052777     0.500032  
min      19.300000     0.000000  
25%      25.888848     0.000000  
50%      28.300000     0.500000  
75%      32.443461     1.000000  
max      37.300000     1.000000

Sau khi xử lý ngoại lai bằng IQR 1.5×, dữ liệu numeric đã hợp lý: age có min=7, max=79 (trước là 0.08–82), trung bình ~55; avg_glucose_level min=58.82, max=132.51, trung bình ~99.86, độ lệch chuẩn giảm từ ~38.8 xuống ~25.45; bmi min=19.3, max=37.3, trung bình ~28.94, độ lệch chuẩn ~5.05, tất cả các giá trị cực đoan đã được giới hạn. Biến nhị phân hypertension (~9%) và heart_disease (~5%) vẫn bất cân bằng, cần lưu ý khi huấn luyện mô hình, trong khi nhãn stroke đã cân bằng hoàn hảo (mean=0.5, min=0, max=1). Tổng thể, dữ liệu numeric giờ sẵn sàng cho mô hình ML, ngoại lai đã được kiểm soát, phân bố hợp lý và nhãn cân bằng. 

In [98]:
stroke_prediction_df_cleand.head(10)

age  hypertension  heart_disease  avg_glucose_level   bmi  gender_Male  \
0  48.0             0              0              69.21  33.1        False   
1  15.0             0              0             122.25  21.0         True   
2  67.0             0              0             110.42  24.9        False   
3  44.0             0              0              65.41  24.8         True   
4  14.0             0              0              82.34  31.6         True   
5  73.0             1              0              71.29  37.3         True   
6  63.0             1              1             132.51  37.3         True   
7  42.0             0              0              75.34  37.3        False   
8  14.0             0              0              70.54  24.4        False   
9  53.0             0              1             132.51  32.0         True   

   gender_Other  ever_married_Yes  work_type_Never_worked  work_type_Private  \
0         False              True                   False               True   
1         False             False                   False               True   
2         False              True                   False              False   
3         False              True                   False               True   
4         False             False                   False              False   
5         False             False                   False              False   
6         False              True                   False               True   
7         False              True                   False              False   
8         False             False                   False               True   
9         False              True                   False               True   

   work_type_Self-employed  work_type_children  Residence_type_Urban  \
0                    False               False                  True   
1                    False               False                 False   
2                     True               False                 False   
3                    False               False                  True   
4                    False               False                  True   
5                    False               False                  True   
6                    False               False                  True   
7                     True               False                 False   
8                    False               False                 False   
9                    False               False                 False   

   smoking_status_formerly smoked  smoking_status_never smoked  \
0                           False                         True   
1                           False                         True   
2                           False                         True   
3                           False                        False   
4                           False                        False   
5                           False                         True   
6                            True                        False   
7                           False                         True   
8                            True                        False   
9                           False                        False   

   smoking_status_smokes  stroke  
0                  False       0  
1                  False       0  
2                  False       0  
3                   True       0  
4                  False       0  
5                  False       0  
6                  False       0  
7                  False       0  
8                  False       0  
9                  False       0